src/VisualStudio/Core/Def/Implementation/EditAndContinue/VsENCRebuildableProjectImpl.cs

```diff
@@ -926,20 +925,42 @@ public unsafe int BuildForEnc(object pUpdatePE)
                 Debug.Assert(_lastEditSessionSummary == ProjectAnalysisSummary.ValidInsignificantChanges ||
                              _lastEditSessionSummary == ProjectAnalysisSummary.ValidChanges);
 
-                var updater = (Interop.IDebugUpdateInMemoryPE2)pUpdatePE;
+                var updater = (IDebugUpdateInMemoryPE2)pUpdatePE;
                 if (_committedBaseline == null)
                 {
-                    Interop.IENCDebugInfo debugInfo;
+                    IENCDebugInfo debugInfo;
                     updater.GetENCDebugInfo(out debugInfo);
 
-                    var symbolReaderProvider = (Interop.IENCSymbolReaderProvider)debugInfo;
-                    symbolReaderProvider.GetSymbolReader(out _pdbReaderMtaPointer);
+                    // ISymUnmanagedReader can only be accessed from an MTA thread, however, we need
+                    // fetch the IUnknown instance (call IENCSymbolReaderProvider.GetSymbolReader) here
+                    // in the STA.  To further complicate things, we need to return synchronously from
+                    // this method.  Waiting for the MTA thread to complete so we can return synchronously
+                    // blocks the STA thread, so we need to make sure the CLR doesn't try to marshal
+                    // ISymUnmanagedReader calls made in an MTA back to the STA for execution (if this
+                    // happens we'll be deadlocked).  We'll use CoMarshalInterThreadInterfaceInStream to
+                    // achieve this.  First, we'll marshal the object in a Stream and pass a pointer to the
+                    // Stream to the MTA.  In the MTA, we'll get the Stream from the pointer and unmarshal
+                    // the object.  The CLR apparently sees this as the IUnknown entering the runtime in the
+                    // MTA, so calls to methods on the symbol reader instance in the MTA will be marshalled
+                    // directly (free threaded calls).  Another way to achieve this would be for the symbol
+                    // reader to implement IAgileObject, but the symbol reader we use today does not.  If
+                    // that changes, we should consider removing the code that marshals the instance using
+                    // CoMarshalInterThreadInterfaceInStream/CoGetInterfaceAndReleaseStream.
+                    var symbolReaderProvider = (IENCSymbolReaderProvider)debugInfo;
+                    object pdbReaderObjSta;
+                    symbolReaderProvider.GetSymbolReader(out pdbReaderObjSta);
+                    int hr = NativeMethods.GetStreamForObject(pdbReaderObjSta, out _pdbReaderObjAsStream);
+                    Marshal.ReleaseComObject(pdbReaderObjSta);
+                    if (hr != VSConstants.S_OK)
+                    {
+                        return VSConstants.E_FAIL;
```

Comment:
The caller expects <= 0, and it appears that's all CoMarshalInterThreadInterfaceInStream returns, so I'll change it to just return hr.